File: channel-core/test/pinstore_field_guards.test.ts
import { describe, expect, test } from "vitest";
import { PinStore } from "../src/identity/pinstore.js";

// These assert the exact production message. The pre-existing coverage of the
// 'pins' mapping check matched /pins.*mapping/i against loadPinStore's wrapped
// error, which embeds the store path — and the temp directory it was given was
// named "pinstore-fc-", so the regex matched the path rather than the guard and
// held even when the guard was gone. Asserting through PinStore.fromYAML keeps
// the path out of the message entirely.
//
// Each fixture is built from a store that loads cleanly
// (TestValidStoreLoads below) and differs in exactly one field, and the failure
// each guard prevents is a store that loads ANYWAY: a scalar in place of the
// pins or addresses mapping yields a silently empty store, which reopens
// first-contact TOFU for every identity the file was holding.

const VALID_PIN = [
  "  did:key:zAlice:",
  "    address: acme.com/alice",
  "    handle: '@alice'",
  "    first_seen: '2026-02-22T10:00:00Z'",
  "    last_seen: '2026-02-22T11:00:00Z'",
];

function store(options: { pins?: string[]; addresses?: string[]; pinExtra?: string[] } = {}): string {
  const pins = options.pins ?? [...VALID_PIN, ...(options.pinExtra ?? []).map((line) => `    ${line}`)];
  const addresses = options.addresses ?? ["  acme.com/alice: did:key:zAlice"];
  return ["pins:", ...pins, "addresses:", ...addresses, ""].join("\n");
}

describe("pin store field guards", () => {
  test("the shared valid store loads", () => {
    const loaded = PinStore.fromYAML(store());
    expect(loaded.pins.get("did:key:zAlice")?.address).toBe("acme.com/alice");
    expect(loaded.addresses.get("acme.com/alice")).toBe("did:key:zAlice");
  });

  // Without this check Object.entries(42) yields nothing and the store loads as
  // a valid EMPTY store — the silent substitution the module exists to prevent.
  test("a scalar in place of the pins mapping is rejected", () => {
    expect(() => PinStore.fromYAML("pins: 42\naddresses: {}\n"))
      .toThrow("pin store 'pins' must be a mapping");
  });

  test("a scalar in place of the addresses mapping is rejected", () => {
    expect(() => PinStore.fromYAML(["pins:", ...VALID_PIN, "addresses: 42", ""].join("\n")))
      .toThrow("pin store 'addresses' must be a mapping");
  });

  test("an empty pin key is rejected", () => {
    expect(() => PinStore.fromYAML(store({
      pins: ["  '':", ...VALID_PIN.slice(1)],
      addresses: [],
    }))).toThrow("pin store has an empty pin key");
  });

  test("an empty address key is rejected", () => {
    expect(() => PinStore.fromYAML(store({ addresses: ["  '': did:key:zAlice"] })))
      .toThrow("pin store has an empty address key");
  });

  // Pin.handle is typed string. A coerced number loads as a number here while
  // Go's requireString refuses it, so the two runtimes would hold different
  // pins for the same file.
  test("a non-string optional field is rejected rather than coerced", () => {
    expect(() => PinStore.fromYAML(store({ pins: [...VALID_PIN.slice(0, 2), "    handle: 42", ...VALID_PIN.slice(3)] })))
      .toThrow("pin 'did:key:zAlice' field 'handle' must be a string");
  });

  // log_seq is the anti-rollback checkpoint. A zero, a negative, a fraction or
  // a quoted digit must not become a checkpoint value.
  test.each(["0", "-1", "1.5", "'3'"])("log_seq %s is rejected", (value) => {
    expect(() => PinStore.fromYAML(store({ pinExtra: [`log_seq: ${value}`] })))
      .toThrow("pin 'did:key:zAlice' field 'log_seq' must be a positive integer");
  });

  test("a positive integer log_seq is accepted", () => {
    expect(PinStore.fromYAML(store({ pinExtra: ["log_seq: 7"] })).pins.get("did:key:zAlice")?.log_seq).toBe(7);
  });
});

File: cli/go/awid/pinstore_field_guards_test.go
package awid

import (
	"strings"
	"testing"
)

// ParsePinStore checks the forward and reverse indexes agree in both directions
// after every per-field guard has run. That check rejects most malformed stores
// on its own, so a fixture whose index does not already agree measures the index
// check rather than the field guard: remove the field guard and the store is
// still refused, just with a different message.
//
// Every fixture here is index-consistent, so removing the guard under test lets
// the store LOAD, which is the failure that matters — a trust database that
// silently accepts a rolled-back checkpoint or a type-coerced field.
// TestConsistentPinStoreLoadsCleanly pins the shared base, and each guard test
// changes exactly one field of it.

type pinFixture struct {
	address   string
	handle    string
	firstSeen string
	lastSeen  string
	extra     []string
}

func validPin() pinFixture {
	return pinFixture{
		address:   "a@b.c",
		handle:    "'@alice'",
		firstSeen: "'2026-02-22T10:00:00Z'",
		lastSeen:  "'2026-02-22T11:00:00Z'",
	}
}

func (f pinFixture) render() string {
	lines := []string{
		"pins:",
		"  did:key:zA:",
		"    address: " + f.address,
		"    handle: " + f.handle,
		"    first_seen: " + f.firstSeen,
		"    last_seen: " + f.lastSeen,
	}
	for _, extra := range f.extra {
		lines = append(lines, "    "+extra)
	}
	lines = append(lines, "addresses:", "  a@b.c: did:key:zA", "")
	return strings.Join(lines, "\n")
}

// rejectsPin asserts the store is refused for the reason named, not by a
// neighbouring check that happens to refuse the same fixture.
func rejectsPin(t *testing.T, f pinFixture, wantSubstring string) {
	t.Helper()

	ps, err := ParsePinStore([]byte(f.render()))
	if err == nil {
		t.Fatalf("store loaded with %d pins, want a load failure: the guard no longer rejects this field", len(ps.Pins))
	}
	if !strings.Contains(err.Error(), wantSubstring) {
		t.Fatalf("error %q does not mention %q: the store was refused, but by a different check", err.Error(), wantSubstring)
	}
}

func TestConsistentPinStoreLoadsCleanly(t *testing.T) {
	t.Parallel()

	ps, err := ParsePinStore([]byte(validPin().render()))
	if err != nil {
		t.Fatalf("ParsePinStore: %v", err)
	}
	if got := ps.Pins["did:key:zA"]; got == nil || got.Address != "a@b.c" {
		t.Fatalf("pin=%+v, want the fixture pin", got)
	}
	if ps.Addresses["a@b.c"] != "did:key:zA" {
		t.Fatalf("addresses=%v, want a consistent reverse index", ps.Addresses)
	}
}

// log_seq is the anti-rollback checkpoint. Without the guard, log_seq: 0 loads
// as an absent checkpoint and the rollback it exists to refuse goes unnoticed.
func TestPinStoreRejectsNonPositiveLogSeq(t *testing.T) {
	t.Parallel()

	f := validPin()
	f.extra = []string{"log_seq: 0"}
	rejectsPin(t, f, "log_seq")

	f.extra = []string{"log_seq: -1"}
	rejectsPin(t, f, "log_seq")
}

// The guard reads the raw node because an explicit 0 and an absent field are
// indistinguishable once decoded; a non-integer must not decode to that same 0.
func TestPinStoreRejectsNonIntegerLogSeq(t *testing.T) {
	t.Parallel()

	f := validPin()
	f.extra = []string{"log_seq: notanumber"}
	rejectsPin(t, f, "log_seq")

	f.extra = []string{"log_seq: '3'"}
	rejectsPin(t, f, "log_seq")
}

// requireString refuses coercion so Go and Node agree on a shared file: yaml.v3
// would happily read 42 as the string "42" where Node's JSON_SCHEMA keeps it a
// number, and the two runtimes would then hold different pins.
func TestPinStoreRejectsCoercedStringField(t *testing.T) {
	t.Parallel()

	f := validPin()
	f.handle = "42"
	rejectsPin(t, f, "must be a string")

	f = validPin()
	f.handle = "true"
	rejectsPin(t, f, "must be a string")
}

func TestPinStoreRejectsEmptyFirstSeen(t *testing.T) {
	t.Parallel()

	f := validPin()
	f.firstSeen = "''"
	rejectsPin(t, f, "first_seen")
}

func TestPinStoreRejectsEmptyLastSeen(t *testing.T) {
	t.Parallel()

	f := validPin()
	f.lastSeen = "''"
	rejectsPin(t, f, "last_seen")
}
